src/Runtime/NetDaemon.Runtime/Internal/AppStateManager.cs

```diff
@@ -1,67 +1,61 @@
-using System.Net;
 using System.Reactive.Linq;
-using System.Text;
 using NetDaemon.AppModel;
-using NetDaemon.Client.HomeAssistant.Model;
-using NetDaemon.Client.Internal.Exceptions;
 
 namespace NetDaemon.Runtime.Internal;
 
 internal class AppStateManager : IAppStateManager, IHandleHomeAssistantAppStateUpdates, IDisposable
 {
+    private readonly IAppStateRepository _appStateRepository;
     private readonly CancellationTokenSource _cancelTokenSource = new();
-    private readonly IServiceProvider _provider;
     private readonly ConcurrentDictionary<string, ApplicationState> _stateCache = new();
 
     public AppStateManager(
-        IServiceProvider provider
+        IAppStateRepository appStateRepository
     )
     {
-        _provider = provider;
+        _appStateRepository = appStateRepository;
     }
 
     public async Task<ApplicationState> GetStateAsync(string applicationId)
     {
-        var entityId = ToSafeHomeAssistantEntityIdFromApplicationId(applicationId);
-        if (_stateCache.TryGetValue(entityId, out var applicationState)) return applicationState;
+        if (_stateCache.TryGetValue(applicationId, out var applicationState)) return applicationState;
 
-        return (await GetOrCreateStateForApp(entityId).ConfigureAwait(false))?.State == "on"
+        return await _appStateRepository.GetOrCreateAsync(applicationId, _cancelTokenSource.Token)
+            .ConfigureAwait(false)
             ? ApplicationState.Enabled
             : ApplicationState.Disabled;
     }
 
     public async Task SaveStateAsync(string applicationId, ApplicationState state)
     {
-        var haConnection = _provider.GetRequiredService<IHomeAssistantConnection>() ??
-                           throw new InvalidOperationException();
-        var entityId = ToSafeHomeAssistantEntityIdFromApplicationId(applicationId);
+        _stateCache[applicationId] = state;
 
-        _stateCache[entityId] = state;
-
-        var currentState = (await GetOrCreateStateForApp(entityId).ConfigureAwait(false))?.State
-                           ?? throw new InvalidOperationException();
+        var isEnabled = await _appStateRepository.GetOrCreateAsync(applicationId, _cancelTokenSource.Token)
+            .ConfigureAwait(false);
 
         switch (state)
         {
-            case ApplicationState.Enabled when currentState == "off":
-                await haConnection.CallServiceAsync("input_boolean", "turn_on",
-                    new HassTarget {EntityIds = new[] {entityId}},
-                    cancelToken: _cancelTokenSource.Token);
-                break;
-            case ApplicationState.Disabled when currentState == "on":
-                await haConnection.CallServiceAsync("input_boolean", "turn_off",
-                    new HassTarget {EntityIds = new[] {entityId}},
-                    cancelToken: _cancelTokenSource.Token);
+            case ApplicationState.Enabled when !isEnabled:
+            case ApplicationState.Disabled when isEnabled:
```

Comment:
```suggestion
            default:
```
It's a boolean so can only have 2 values?

Reply:
the state is an enum that can be Enabled, disabled or some other states I think. But I agree the switch here could probably better be replaces with a normal if

Reply:
lol, yeah I was about to comment 2 things;
1) This should probably be just a if
2) @ludeeus you have no clue what you are talking about 
:smile:

Reply:
Haha